scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,247 @@
+      timeline.add(delta);
+      // TODO: Advance a dense time counter so that future tasks are strictly ordered relative to these,
+      //   even if they occur at the same real time.
+
+      // Run the jobs in this batch.
+      final var commit = engine.performJobs(batch.jobs(), cells, curTime, Duration.MAX_VALUE, missionModel);
+      timeline.add(commit);
+    }
+    areLastSimResultsDirty = true;
+  }
+
+  private void simulateManyActs(ArrayList<SimulatedActivity> acts){
+    final var controlTask = buildControlTask(acts);
+    simulateControlTask(controlTask);
+    activitiesInserted.addAll(acts);
+  }
+
+  private void simulateAct(SimulatedActivity simAct){
+    final var controlTask = buildControlTask(simAct);
+    simulateControlTask(controlTask);
+    activitiesInserted.add(simAct);
+  }
+
+  /**
+   * Returns the duration of a terminated simulated activity
+   * @param actName the activity name
+   * @return its duration if the activity has been simulated and has finished simulating, an IllegalArgumentException otherwise
+   */
+  public Duration getActivityDuration(String actName){
+    return engine.getTaskDuration(actToId.get(actName));
+  }
+
+  private final class ControlTask implements Task {
+    private final Map<String, Pair<Duration, SerializedActivity>> schedule;
+
+    /* The directive that caused a task (if any). */
+    // Non-final because we replace it with an empty map when extracted by a client.
+    private Map<String, String> taskToPlannedDirective = new HashMap<>();
+
+    private final PriorityQueue<Triple<Duration, String, SerializedActivity>> scheduledTasks
+        = new PriorityQueue<>(Comparator.comparing(Triple::getLeft));
+
+    private Duration currentTime;
+
+    public ControlTask(final Map<String, Pair<Duration, SerializedActivity>> schedule, Duration curTime) {
+      this.schedule = Objects.requireNonNull(schedule);
+      this.currentTime = curTime;
+      this.reset();
+    }
+
+    public Map<String, String> extractTaskToPlannedDirective() {
+      final var taskToPlannedDirective = this.taskToPlannedDirective;
+      this.taskToPlannedDirective = new HashMap<>();
+      return taskToPlannedDirective;
+    }
+
+    @Override
+    public TaskStatus step(final Scheduler scheduler) {
+      while (true) {
+        var nextTask = this.scheduledTasks.peek();
+        if (nextTask == null) break;
+
+        final var startTime = nextTask.getLeft();
+        if (startTime.longerThan(this.currentTime)) {
+          final var delta = nextTask.getLeft().minus(this.currentTime);
+          this.currentTime = nextTask.getLeft();
+          return TaskStatus.delayed(delta);
+        }
+
+        this.scheduledTasks.remove();
+
+        final var directiveId = nextTask.getMiddle();
+        final var specification = nextTask.getRight();
+
+        final var id = scheduler.spawn(specification.getTypeName(), specification.getParameters());
+        actToId.put(nextTask.getMiddle(), id);
```

Comment:
`getMiddle()` is already saved as a (more helpfully-named) variable, `directiveId`:
```suggestion
        actToId.put(directiveId, id);
```
In particular, this makes it much more obvious that `actToId` is the inverse of `taskToPlannedDirective`, which suggests we might want to rename it to `plannedDirectiveToTask` instead.

Reply:
I'll replace the `getMiddle()` by `directiveId` but I find that `plannedDirective` makes little sense in the context of the incremental simulation (`actToId` is part of the `IncrementalSimulationDriver`), when it is really activity ids (coming originally from the scheduler) to task id (issued by the engine), `actIdToTaskId`?

Reply:
Whatever terminology is settled on, I'd mostly just like the `xToY` and `yToX` naming to be upheld. That means you might want/need to rename the existing `taskToPlannedDirective` map to match your intended convention.

Reply:
Also, the difference between "activity" and "directive" is simply that "directive" applies to both the activity modeling and sequence/command modeling domains, while "activity" is particular. Since this class is owned by activity-specific scheduling domain, I don't see a problem with using "activity", but I do think the `IncrementalSimulationEngine` could very easily be lifted out of the scheduler in the future -- it's not truly activity-specific on its own, it just caters to use-cases surfaced by the scheduler.

Reply:
(As a concrete suggestion, I'd be in favor of "activityToTask" and "taskToActivity". I tend to be less a fan of shortened words like "act", although it's unambiguous in this case.)

Reply:
In the latest commit I have actually went back to the original directive/task naming and let reference to activities in a facade for the scheduler.